Coding standards/PHP coding standards/Functional programming.md

```diff
@@ -0,0 +1,209 @@
+
+Functions are first-class citizens in PHP:
+
+```php
+<?php
+// Function as a variable.
+$function_name = function() {
+  return 42;
+};
+
+// Function as a return type.
+function create_function() {
+  return function() {
+    return 'Return from an anonymous function.';
+  };
+};
+
+$new_func = create_function();
+
+// Function as a parameter.
+function display( $function ) {
+  echo $func() , "\n\n";
+}
+
+// Call a function by name.
+call_user_func_array( 'strtoupper', $some_string );
+
+// objects as functions.
+class Some_Class {
+  public function __invoke( $param1, $param2 ) {
+    [...]
+  }
+}
+
+$instance = new Some_Class();
+$instance( 'First', 'Second' ); // call the __invoke() method
+```
+
+Even though looping through iterable objects is generally faster if you use `for` or `foreach`, you can use some functional programming techniques that will make your code a bit less _boilerplatey_. Also, the performance effects are [negligible](https://stackoverflow.com/questions/18144782/performance-of-foreach-array-map-with-lambda-and-array-map-with-static-function).
+
+### Mapping
+
+Applying a function to all elements:
+
+```php
+<?php
+// Old way (imperative):
+foreach ( $iterable as $iterable_key => $iterable_value ) {
+  $iterable[ $iterable_key ] = strtoupper( $iterable_value );
+}
+
+// New way (functional):
+$result = array_map( 'strtoupper', $iterable );
+```
+
+[array_map manual](https://php.net/manual/en/function.array-map.php)
+
+### Reducing
+
+Reduce an array to a single value:
+
+```php
+<?php
+// Old way (imperative):
+$result = 0;
+
+foreach ( $int_array as $value ) {
+  $result += $value;
+}
+
+// New way (functional):
+$result = array_reduce( $int_array, function( $carry, $item ) { return $carry + $item; }, 0 );
+```
+
+[array_reduce manual](https://php.net/manual/en/function.array-reduce.php)
+
+### Filtering
+
+Iterating over an array, but returning only those results that pass some conditions:
+
+[array_filter manual](https://php.net/manual/en/function.array-filter.php)
+
+```php
+<?php
+// Old way (imperative):
+$result = [];
+
+foreach ( $int_array as $value ) {
+  if( $value % 2 === 0 ) {
+    $result[] = $value;
+  }
+}
+
+// New way (functional):
+$result = array_filter( $int_array, function( $item ) { return ( $item % 2 === 0); } );
+```
+
+### Anonymous functions
+
+Using anonymous functions for actions and filters could be problematic because it makes them very hard to unhook later on.
+
+```php
+<?php
+add_action( 'init', function() {
+  call_function();
+}, 10 );
+```
+
+Do this instead:
+
+```php
+<?php
+add_action( 'init', 'my_callable_function', 10 );
+
+function my_callable_function() {
+  call_function();
+}
+```
+
+If you are working on a 'public' project that can be extended, you should follow the above instructions. When working on client projects, these are not that important, as you don't expect anybody to remove your added hooks (in themes or in plugins).
```

Comment:
> When working on client projects, these are not that important, as you don't expect anybody to remove your added hooks (in themes or in plugins).

This sentence sounds a bit wrong. Maybe something like this would sound better:
When working on client projects, there is a smaller chance that your hooks will be removed. In that case, Anonymous functions are not that important.